Weapon Ammo Refill Fix

PYROTEKNIK opened the pull request:
WeaponAndAmmoProduct items should refill ammo correctly now. It will use game-standard ammo refill behavior when no data is supplied.

In most cases the ammotype field can remain blank, as it is acquired automatically. amount will attempt to get SWEP.Primary.DefaultClip, or weapon:GetMaxClip1() if not specified. You should probably manually enter the amount to give for HL2 weapons, otherwise you get whatever a single clip can hold (nothing if if it's a clipless weapon like the grenade, rpg, or crossbow)

maxammo can be used to limit how much ammo a player can carry at once. 


Make sure if you're using WeaponAndAmmoProduct, you actually set up ammo properly on your swep.

PYROTEKNIK committed:
Update sh_products.lua

WeaponAndAmmoProduct items should refill ammo correctly now. It will use game-standard ammo refill behavior when no data is supplied.

In most cases the ammotype field can remain blank, as it is acquired automatically. amount will attempt to get SWEP.Primary.DefaultClip, or weapon:GetMaxClip1() if not specified. You should probably manually enter the amount to give for HL2 weapons, otherwise you get whatever a single clip can hold (nothing if if it's a clipless weapon like the grenade, rpg, or crossbow)

Make sure if you're using it, you actually set up ammo on your swep.

PYROTEKNIK committed:
Update sh_products.lua

Edited to factor in clipsize 0 as valid for non-clip weapons

swampservers commented on lua/swampshop/sh_products.lua:
```diff
@@ -83,28 +83,52 @@ function SS_WeaponProduct(product)
         self:SS_WeaponProduct_OnBuy(ply)
     end
 
+    function product:CannotBuy(ply)
```
Can you make it select the weapon? So a bind can either buy the weapon, or just select it

PYROTEKNIK replied:
Will it still charge the player?

swampservers replied:
Select weapon then return a string, and it won't buy it

PYROTEKNIK replied:
oh okay

swampservers commented on lua/swampshop/sh_products.lua:
```diff
@@ -83,28 +83,52 @@ function SS_WeaponProduct(product)
+            local wep = ply:GetWeapon(self.class)
+            local ammotype = self.ammotype or (game.GetAmmoName(wep:GetPrimaryAmmoType()))
+            local ammogive = self.amount or (wep.Primary and wep.Primary.DefaultClip) or wep:GetMaxClip1() or 0
+            if (ammogive <= 0) then
```
Why not just check ammogive <= 0 in the if statement

swampservers commented on lua/swampshop/sh_products.lua:
```diff
@@ -83,28 +83,52 @@ function SS_WeaponProduct(product)
+        if (ply:HasWeapon(self.class)) then
+            local wep = ply:GetWeapon(self.class)
+            local ammotype = self.ammotype or (game.GetAmmoName(wep:GetPrimaryAmmoType()))
+            local ammogive = self.amount or (wep.Primary and wep.Primary.DefaultClip) or wep:GetMaxClip1() or 0
```
Does :GetMaxClip1() return the same thing as (wep.Primary and wep.Primary.DefaultClip) for sweps?

PYROTEKNIK replied:
I'm honestly not sure, I got inconsistent results, but defaultclip is more reliable, since it's a user editable var

swampservers commented on lua/swampshop/sh_products.lua:
```diff
@@ -83,28 +83,52 @@ function SS_WeaponProduct(product)
+            if (ammogive <= 0) then
+                ammogive = nil
+            end
+            if (ammotype == nil or ammogive == nil) then return "Cannot Refill Ammo. Please report this as a bug." end
```
assert()

PYROTEKNIK committed:
Update sh_products.lua

Not calling OnBuy should stop the player from being charged points, right?
Assert

PYROTEKNIK committed:
don't charge player if they try to buy something while dead

swampservers commented on lua/swampshop/sh_products.lua:
```diff
@@ -78,33 +78,55 @@ function SS_WeaponProduct(product)
+            local wep = ply:GetWeapon(self.class)
+            local ammotype = self.ammotype or (game.GetAmmoName(wep:GetPrimaryAmmoType()))
+            local ammogive = self.amount or (wep.Primary and wep.Primary.DefaultClip) or wep:GetMaxClip1() or 0
+            local fail = assert(ammotype != nil and ammogive > 0,"shop purchase could not refill" .. ply:Nick() .. "'s "..self.class.." with ".. ammogive .." of "..(ammotype or "<nil ammotype>") ,true)
```
This isn't what assert does...? https://wiki.facepunch.com/gmod/Global.assert

PYROTEKNIK replied:
I looked at this and I don't entirely understand what it returns.

swampservers commented:
handled in https://github.com/swampservers/contrib/pull/268/ thanks
